src/Type.h

```diff
@@ -734,9 +737,20 @@ class RecordType final : public Type
 
 	detail::TraversalCode Traverse(detail::TraversalCallback* cb) const override;
 
+	// Can initialization of record values of this type be deferred?
+	//
+	// When record types contain non-const &default expressions or recursively
+	// contain any nested records that themselves are not deferrable,
+	// initialization can not be deferred, otherwise possible.
+	bool IsDeferrable() const;
+
+	static void DoneParsing();
+
 private:
 	RecordType() { types = nullptr; }
 
+	static std::vector<zeek::RecordTypePtr> record_types;
```

Comment:
Instead of adding more hidden global state, suggest to use AST traversal instead, that's what's it's meant for: examining language elements after parsing.

Reply:
Yeah, this is nicer. As in DM, requires a `const_cast`.

I made it a private class on RecordType so it can work directly with `creation_inits` of individual instances.